### PR TITLE
persist-txn: add dyncfg to apply_eager on init instead of forget_all

### PR DESCRIPTION
**Pull request by danhhz**
In #25992, we discovered that persist-txn forget records do not
consolidate out of the raw shard as intended (they work fine in-mem).
This combined with the forget_all call as part of boot creates a risk
that a crashlooping envd process will quickly increase the physical size
of the txns shard.

Luckily, the forget_all call was only needed as part of the story for
rollback back to the old tables implementation, which has since been
removed. We can now get away with a lesser set of invariants at boot.
Put this behind a dyncfg because it's occasionally useful to have the
forget_all.

Mechanically, I add the dyncfg with a default of the previous behavior
in this commit. In the next, I'll switch the default. This allows us to
potentially cherry-pick just this commit into the currently qualifying
0.92.x release to get the knob out sooner.

Touches #25992

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/persist-txn/src/lib.rs [hunk 1]
```diff
@@ -265,11 +265,19 @@ mod proto {
...
+        .add(&crate::operator::DATA_SHARD_RETRYER_CLAMP)
         .add(&crate::operator::DATA_SHARD_RETRYER_INITIAL_BACKOFF)
         .add(&crate::operator::DATA_SHARD_RETRYER_MULTIPLIER)
-        .add(&crate::operator::DATA_SHARD_RETRYER_CLAMP)
 }
 
+/// Used by storage controller as a temporary escape hatch for #25992.
+pub const INIT_FORGET_ALL: Config<bool> = Config::new(
+    "persist_txn_init_forget_all",
+    false,
```
Comment (jkosh44) [inline src/persist-txn/src/lib.rs:277]: Shouldn't this be true so that the default behavior is that we call `forget_all`?

**Review by jkosh44 — APPROVED** [review]

File: src/persist-txn/src/lib.rs (hunk 1 above)
Comment (danhhz) [reply src/persist-txn/src/lib.rs:277]: I was thinking that we don't strictly need the forget_all anymore (and with the bug it's accumulating things each week), plus if some envd got into a crash loop, we'd have to react by flipping the LD. Open to flipping it though!

(NB: The second commit changes the default from true->false, and if we cherry-picked this, we'd only do the first one. If that changes things for you.)

Comment (jkosh44) [reply src/persist-txn/src/lib.rs:277]: Oh got it, I misunderstood and didn't realize there were two commits. This LGTM as is.

**Comment by danhhz** [comment]
Nightlies are green, too. TFTR!

**Pull request merged**